src/league_gui.cpp

```diff
@@ -0,0 +1,453 @@
+/*
+ * This file is part of OpenTTD.
+ * OpenTTD is free software; you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 2.
+ * OpenTTD is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details. You should have received a copy of the GNU General Public License along with OpenTTD. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file league_gui.cpp GUI for league tables. */
+
+#include "stdafx.h"
+#include "league_gui.h"
+
+#include "company_base.h"
+#include "company_gui.h"
+#include "gui.h"
+#include "industry.h"
+#include "league_base.h"
+#include "sortlist_type.h"
+#include "story_base.h"
+#include "strings_func.h"
+#include "tile_map.h"
+#include "town.h"
+#include "viewport_func.h"
+#include "window_gui.h"
+#include "zoom_func.h"
+#include "widgets/league_widget.h"
+#include "table/strings.h"
+#include "table/sprites.h"
+
+#include "safeguards.h"
+
+
+static const StringID _performance_titles[] = {
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_ENGINEER,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_ENGINEER,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_TRAFFIC_MANAGER,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_TRAFFIC_MANAGER,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_TRANSPORT_COORDINATOR,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_TRANSPORT_COORDINATOR,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_ROUTE_SUPERVISOR,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_ROUTE_SUPERVISOR,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_DIRECTOR,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_DIRECTOR,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_CHIEF_EXECUTIVE,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_CHIEF_EXECUTIVE,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_CHAIRMAN,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_CHAIRMAN,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_PRESIDENT,
+	STR_COMPANY_LEAGUE_PERFORMANCE_TITLE_TYCOON,
+};
+
+static inline StringID GetPerformanceTitleFromValue(uint value)
+{
+	return _performance_titles[std::min(value, 1000u) >> 6];
+}
+
+class PerformanceLeagueWindow : public Window {
+private:
+	GUIList<const Company*> companies;
```

Comment:
As I understand it, because it's inconvenient to keep list constantly updated, so to avoid glitches due to resorting it stores a copy and updates it episodically with full redraw of the window.